### ISSUE TITLE
autocomplete does not forget 'whatever *'

### ISSUE THREAD

**Issue by The-Alchemist** [fish-shell/fish-shell#2019]
This might be a feature instead of a bug, but it's worth a try to ask:

```
# ls *
# ls &
# ls [*]     <-----  brackets mean autocomplete
```

The autocomplete if you type in ```ls``` doesn't show ```ls &```, as one would expect because it's the most recent command that starts with ```ls```.

It shows ```ls *``` instead.  Is this correct behavior?

NOTE: I'm just using ```ls``` as an illustration.  I keep bumping into this with much more complex commands, where I expect the autocomplete to show me the command as it ends with ```&``` instead of ```*````.

**Comment by ridiculousfish** [comment]
Hmm, I can't reproduce this on OS X Yosemite 10.10.2, with ToT. What version of fish are you using, on what OS?

**Comment by zanchey** [comment]
I can't reproduce it with the git master, but I can with 2.1.1.

**Comment by The-Alchemist** [comment]
```
$ fish --version
fish, version 2.1.2
```
Yosemite 10.10.2.

However, what may be complicating the issue is that I've been using fish for >6 months now, so my history stretches that far.  I probably started with 2.1.0 and upgraded from there, keeping my history in-tact.

I really don't want to ``history --clear``, but I did a ```history --delete --prefix ....``` and selectively deleted some entries, and it seems like the issue still exists.

**Comment by ridiculousfish** [comment]
Fix bisects to c6eef166646b17e409dda043dee311ad074ff4c9

**Comment by ridiculousfish** [comment]
This is due to a bug in the original `autosuggest_parse_command`. If a command ends with a statement terminator (`;` or `&`) we returned false, meaning that we didn't have a command (since we expected another one). This was fixed with the new parser, where we can just find the last command directly.

**Comment by ridiculousfish** [comment]
I think the fix is well understood and is in master. Closing. Thanks for reporting this!

**Comment by The-Alchemist** [comment]
@ridiculousfish : That makes sense, thank you for the fix.

But I'm confused, because it seems like the fix should be in 2.1.2, but that's what I'm using to reproduce this issue.

Am I mis-understanding the issue?

**Comment by snnw** [comment]
@The-Alchemist: 2.1.2 was spun from a separate branch and is missing a lot from master, including c6eef16.  You can try running Nightly builds, if you want the latest features and bugfixes :)
